extensions/typescript/src/features/jsDocCompletionProvider.ts

```diff
@@ -0,0 +1,184 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import { Position, Selection, Range, CompletionItemProvider, CompletionItemKind, TextDocument, CancellationToken, CompletionItem, window, commands, Uri, ProviderResult, TextEditor } from 'vscode';
+
+import { ITypescriptServiceClient } from '../typescriptService';
+import { FileLocationRequestArgs, DocCommandTemplateResponse } from '../protocol';
+
+import * as nls from 'vscode-nls';
+const localize = nls.loadMessageBundle();
+
+const tryCompleteJsDocCommand = '_typeScript.tryCompleteJsDoc';
+
+
+class JsDocCompletionItem extends CompletionItem {
+	constructor(file: Uri, position: Position) {
+		super('/** @param */', CompletionItemKind.Snippet);
+		this.detail = localize('typescript.jsDocCompletionItem.detail', 'Complete JSDoc comment');
+		this.insertText = '';
+		this.command = {
+			title: 'Try Complete Js Doc',
+			command: tryCompleteJsDocCommand,
+			arguments: [file, position]
+		};
```

Comment:
The TypeScript API that we are using for completions is pretty finicky. It will not work if we have a partial comment (such as `/** function abc() {}` ) in the document, so we have to remove the existing comment before making the request against TypeScript for the JSDoc template.

I previously implemented this as you described using insert text and fetching the doc template in `resolveCompletionItem`, but the behavior turned out to be pretty poor since we still have to edit the document to remove the partial comment before making the request. Using a command seems like the best behavior currently.

For the next iteration, we have a few options:

* TS could be smarter about returning JSDoc templates when you already have started typing out the start of a comment
* We could investigate updating a virtual document instead of the actual document in order to make the doc template request